src/python/pants/backend/openapi/util_rules/generator_process.py

```diff
@@ -0,0 +1,137 @@
+from pants.jvm.util_rules import rules as jvm_util_rules
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.meta import frozen_after_init
+
+
+@unique
+class OpenAPIGeneratorType(Enum):
+    JAVA = "java"
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class OpenAPIGeneratorProcess:
+    argv: tuple[str, ...]
+    generator_type: OpenAPIGeneratorType
+    input_digest: Digest
+    description: str = dataclasses.field(compare=False)
+    level: LogLevel
+    extra_env: FrozenDict[str, str]
+    extra_immutable_input_digests: FrozenDict[str, Digest]
+    extra_classpath_entries: tuple[str, ...]
+    extra_jvm_options: tuple[str, ...]
+    cache_scope: ProcessCacheScope | None
+    output_directories: tuple[str, ...]
+    output_files: tuple[str, ...]
+
+    def __init__(
+        self,
+        *,
+        generator_type: OpenAPIGeneratorType,
+        argv: Iterable[str],
+        input_digest: Digest,
+        description: str,
+        level: LogLevel = LogLevel.INFO,
+        output_directories: Iterable[str] | None = None,
+        output_files: Iterable[str] | None = None,
+        extra_env: Mapping[str, str] | None = None,
+        extra_classpath_entries: Iterable[str] | None = None,
+        extra_immutable_input_digests: Mapping[str, Digest] | None = None,
+        extra_jvm_options: Iterable[str] | None = None,
+        cache_scope: ProcessCacheScope | None = None,
+    ):
+        self.generator_type = generator_type
+        self.argv = tuple(argv)
+        self.input_digest = input_digest
+        self.description = description
+        self.level = level
+        self.output_directories = tuple(output_directories or ())
+        self.output_files = tuple(output_files or ())
+        self.extra_env = FrozenDict(extra_env or {})
+        self.extra_classpath_entries = tuple(extra_classpath_entries or ())
+        self.extra_immutable_input_digests = FrozenDict(extra_immutable_input_digests or {})
+        self.extra_jvm_options = tuple(extra_jvm_options or ())
+        self.cache_scope = cache_scope
+
+
+_GENERATOR_CLASS_NAME = "org.openapitools.codegen.OpenAPIGenerator"
+
+
+@rule
+async def openapi_generator_process(
+    request: OpenAPIGeneratorProcess, jdk: InternalJdk, subsystem: OpenAPIGenerator
+) -> Process:
+    lockfile_request = await Get(GenerateJvmLockfileFromTool, OpenAPIGeneratorLockfileSentinel())
+    tool_classpath = await Get(ToolClasspath, ToolClasspathRequest(lockfile=lockfile_request))
+
+    toolcp_relpath = "__toolcp"
+    immutable_input_digests = {
+        toolcp_relpath: tool_classpath.digest,
+        **request.extra_immutable_input_digests,
+    }
+
+    classpath_entries = [
+        *tool_classpath.classpath_entries(toolcp_relpath),
+        *request.extra_classpath_entries,
+    ]
+
+    extra_jvm_options = [*subsystem.jvm_options, *request.extra_jvm_options]
+
+    jvm_process = JvmProcess(
```

Comment:
Since this code is specific to the JVM, it's probably not generic to all of `openapi` the way its module namespace might suggest.

But also: since this is used in exactly one other `@rule`, it feels like premature optimization to extract it?

Reply:
it is indeed a bit of a premature optimisation. The idea behind this is to make it easier to implement code generation for other languages as all of them use the same generation tool.

In other words: this is not specific to the Java code generator, the code generation tool is a JVM tool which can generate code for a wide variety of languages as it takes in its argument list the target language (aka _generator_) to be used: https://openapi-generator.tech/docs/generators

So with this rule, supporting new generators consist on adding a new enum value to `OpenAPIGeneratorType` and then implementing the backend for that generator so that it uses this rule to build the final process the same way the `java` generator does.

@jyggen mentioned he was interested in implementing the Python code generation, which would benefit from this. I guess the main inconvenience is that this will in fact still depend on the JVM rules.

Reply:
Ok: sounds good. Thanks!